include/graphblas/reference/blas3.hpp

```diff
@@ -1247,134 +1237,91 @@ namespace grb {
 #endif
 			if( m != m_mask || n != n_mask ) {
 #ifdef _DEBUG
-				std::cout << "Mask dimensions do not match input matrix dimensions\n" << std::flush;
+				std::cerr << "Mask dimensions do not match input matrix dimensions\n" << std::flush;
 #endif
-				return RC::MISMATCH;
+				return MISMATCH;
 			}
 
 			RC rc = SUCCESS;
 			RC local_rc = rc;
 			auto local_x = identity;
+
 #ifdef _H_GRB_REFERENCE_OMP_BLAS3
 	#pragma omp parallel default(none) \
 			shared(A_raw, mask_raw, x, rc, std::cout) \
-			firstprivate(local_x, local_rc, m, n, op, identity)
+			firstprivate(local_x, local_rc, m, n, op, identity, mask_identity)
 #endif
 			{
 				size_t start_row = 0;
 				size_t end_row = m;
 #ifdef _H_GRB_REFERENCE_OMP_BLAS3
 				config::OMP::localRange( start_row, end_row, 0, m );
 #endif
+				bool local_visited_columns[ n ] = { false };
```

Comment:
Instead of this one, which performs dynamic allocation on the fly, the suggested approach is to use a pre-allocated thread-local SPA. In the (sequential) reference backend, this is done via `internal::getMatrixBuffers`, as exemplified in `reference/blas3.hpp`. Those buffers are then used to create a temporary `internal::Coordinates` class, to play the role of `local_visited_columns`.

The difference with this code is, however, that we want to run in parallel and that `local_visited_columns` is a thread-local array. This leads to a scalability issue: if there are `T` threads each with an array of size `n`, then `T*n` may well be much larger than `nz`, where `nz` is the number of nonzeroes the primitive is to process-- meaning, calling this primitive would require a much larger work space than the operands of the primitive would require. This does not scale, and therefore in this parallelisation strategy, `T` must be limited to something that is scalable (e.g., `T=floor(nz/n)`).

ALP does guarantee at least one buffer (through `getMatrixBuffers`). It also does guarantee a global buffer space that is proportional to the largest `nz` a container has ever been constructed with. This buffer can be retrieved via calls to
```
const size_t nbytes = internal::template getCurrentBufferSize< char >();
char * buffer = internal::template getReferenceBuffer< char >( nbytes );
```
(The above example retrieves it as a raw byte buffer. In the current context, `nbytes` could be used to derive how many `local_visited_columns` are supported by the current global buffer, and select `T` accordingly. The reference backends during the construction of vector and matrix containers and during any call to `grb::resize` ensures that the global buffer is always of size `Theta( max{ m, n, nz } )`.)

The motivation here are performance semantics -- we would like ALP primitives to execute without performing any system calls, which entails pre-allocating buffers like these ones, without incurring unscalable behaviour.

Reply:
Very clear answer thank you.

I tried to go another way without any additional allocation and threading, not sure about it though. 
See the masked variant of reference/blas3.hpp::fold_masked_generic @anyzelman